overlord/devicestate/devicestate.go

```diff
@@ -294,22 +294,21 @@ func CanManageRefreshes(st *state.State) bool {
 }
 
 // extractDownloadInstallEdgesFromTs is a helper that extract the first, last
-// download and install tasks from a TaskSet
+// download phase and install phase tasks from a TaskSet
 func extractDownloadInstallEdgesFromTs(ts *state.TaskSet) (firstDl, lastDl, firstInst, lastInst *state.Task) {
 	edgeTask := ts.Edge(snapstate.DownloadAndChecksDoneEdge)
-	for _, t := range ts.Tasks() {
-		if firstDl == nil {
-			firstDl = t
-		}
+	tasks := ts.Tasks()
+	// we know we always start with downloads
+	firstDl = tasks[0]
+	// and always end with installs
+	lastInst = tasks[len(tasks)-1]
+	for _, t := range tasks {
 		if firstInst == nil && lastDl != nil {
 			firstInst = t
 		}
 		if t == edgeTask {
 			lastDl = t
```

Comment:
nitpick: this could be a bool flag, and then we hard code edgeTask in the `return`